lib/extensions/permissions.js

```diff
@@ -215,16 +220,84 @@ async function setAccess (bundleId, permissionsMapping) {
     const permissionsArg = _.toPairs(wixPermissions)
       .map((x) => `${x[0]}=${formatStatus(x[1])}`)
       .join(',');
-    await execWix.bind(this)([
+    const execWixFn = async () => await execWix.bind(this)([
       '--byId', this.udid,
       '--bundle', bundleId,
       '--setPermissions', permissionsArg,
     ]);
+    const shouldWaitForSystemReadiness = !_.isEmpty(
+      _.intersection(SERVICES_NEED_SPRINGBOARD_RESTART, _.keys(wixPermissions))
+    );
+    if (shouldWaitForSystemReadiness) {
+      const [didTimeout] = await runAndWaitForSystemReadiness.bind(this)(
+        execWixFn, SYSTEM_SERVICE_RESTART_TIMEOUT_MS
+      );
+      if (didTimeout) {
+        this.log.warn(
+          `The Springboard service and apps restarted by the springboard process restart did not restart ` +
+          `after ${SYSTEM_SERVICE_RESTART_TIMEOUT_MS}ms timeout. This might lead to unexpected consequences later.`
+        );
+      }
+    } else {
+      await execWixFn();
+    }
   }
 
   return true;
 }
 
+/**
+ * Waiting for springboard restart and applications process end/restart
+ * triggered by the springboard process restart.
+ *
+ * @this {import('../types').CoreSimulator}
+ * @template {any} T
+ * @param {() => Promise<T>} fn
+ * @param {number} timeoutMs
+ * @returns {Promise<[boolean, T]>}
+ */
+async function runAndWaitForSystemReadiness(fn, timeoutMs) {
+  const waitForNewPid = async (initialPid, bundleId, timeoutMs) => {
+    await waitForCondition(async () => {
+      try {
+        const pid = (await this.ps()).find(({name}) => bundleId === name)?.pid;
+        return _.isInteger(pid) && initialPid !== pid;
+      } catch {
+        return false;
+      }
+    }, {waitMs: timeoutMs, intervalMs: 500});
+  };
+
+  let initialProcesses = [];
+  try {
+    initialProcesses = await this.ps();
+  } catch {};
+
+  const [initialSpringboardPid, initialSpotlightPid] = [
+    SPRINGBOARD_BUNDLE_ID, SPOTLIGHT_BUNDLE_ID
+  ].map((bundleId) => initialProcesses.find(({name}) => bundleId === name)?.pid);
+
+  const result = await fn();
+  if (!_.isInteger(initialSpringboardPid) || !_.isInteger(initialSpotlightPid)) {
+    // there is no point to wait if relevant processes were not running before
+    return [false, result];
+  }
+
+  try {
+    // Make sure the springboard process restarted first.
+    const timer = new timing.Timer().start();
+    await waitForNewPid(initialSpringboardPid, SPRINGBOARD_BUNDLE_ID, timeoutMs);
+    const remainingTimeout = timeoutMs - timer.getDuration().asMilliSeconds;
+
+    // Then, checking if the new spring board process refreshes applications.
+    // Spotlight.app is widely used so the app process can be an indicator to check the refresh.
+    await waitForNewPid(initialSpotlightPid, SPOTLIGHT_BUNDLE_ID, remainingTimeout);
```

Comment:
we may shortcircuit here if remainingTimeout <= 0

Reply:
done. actially this case could be rare as previous wait for the thing goes to catch clause then but it's reasonable to add the condition check